src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -0,0 +1,154 @@
+            }
+
+            _logger = loggerFactory.CreateLogger<CSharpFormattingPass>();
+        }
+
+        // Run after the C# formatter pass.
+        public override int Order => DefaultOrder - 3;
+
+        public override bool IsValidationPass => false;
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result, CancellationToken cancellationToken)
+        {
+            if (context.IsFormatOnType)
+            {
+                // We don't want to handle OnTypeFormatting here.
+                return result;
+            }
+
+            // Apply previous edits if any.
+            var originalText = context.SourceText;
+            var changedText = originalText;
+            var changedContext = context;
+            if (result.Edits.Length > 0)
+            {
+                var changes = result.Edits.Select(e => e.AsTextChange(originalText)).ToArray();
+                changedText = changedText.WithChanges(changes);
+                changedContext = await context.WithTextAsync(changedText);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Format the razor bits of the file
+            var syntaxTree = changedContext.CodeDocument.GetSyntaxTree();
+            var edits = FormatRazor(changedContext, syntaxTree);
+
+            // Compute the final combined set of edits
+            var formattingChanges = edits.Select(e => e.AsTextChange(changedText));
+            changedText = changedText.WithChanges(formattingChanges);
+            var finalChanges = SourceTextDiffer.GetMinimalTextChanges(originalText, changedText, lineDiffOnly: false);
+            var finalEdits = finalChanges.Select(f => f.AsTextEdit(originalText)).ToArray();
+
+            return new FormattingResult(finalEdits);
+        }
+
+        private IEnumerable<TextEdit> FormatRazor(FormattingContext context, RazorSyntaxTree syntaxTree)
+        {
+            var edits = new List<TextEdit>();
+            var source = syntaxTree.Source;
+
+            foreach (var node in syntaxTree.Root.DescendantNodes())
+            {
+                // Disclaimer: CSharpCodeBlockSyntax is used a _lot_ in razor so I'm being overly careful to only try to
+                // format syntax forms we care about.
+                //
+                // We're looking for a code block like this:
+                //
+                // @code {
+                //    var x = 1;
+                // }
+                //
+                // The nodes will be a grandchild of a RazorDirective (the "@code") and we expect there to be
+                // at least three children, being:
+                // 1. Optional whitespace
+                // 2. The opening brace
+                // 3. The C# code
+                // 4. The closing brace
+                if (node is CSharpCodeBlockSyntax code &&
+                    node.Parent?.Parent is RazorDirectiveSyntax directive &&
```

Comment:
I wonder if its worth bothering to run the pass at all if there are diagnostics?

In any case though, any tips on how to create a malformed directive that still parses correctly, that it would match this?
I tested with `@david   {  }` and `@code foo {  }`, or `@code   }` but the tree is completely different in those cases, and this if statement is sufficiently specific. That's _why_ it's so specific in fact, rather than just looking for code blocks.

Reply:
> I wonder if its worth bothering to run the pass at all if there are diagnostics?

Interesting idea. I think having some level of formatting would probably still be valuable. Aka, only format leaf nodes that have no diagnostics anywhere in their hierarchy.



> In any case though, any tips on how to create a malformed directive that still parses correctly, that it would match this?
> I tested with `@david { }` and `@code foo { }`, or `@code }` but the tree is completely different in those cases, and this if statement is sufficiently specific. That's _why_ it's so specific in fact, rather than just looking for code blocks.

Try out `@code {` (no end brace)

Reply:
Thanks for the suggestion, that still doesn't match the condition here though, because the code block doesn't have 4 children 😛

Then again, there is probably an argument to be made that even if there is a missing end brace, we could still format the start brace?  Either way, I'll throw in the diagnostics condition just for fun. Right now the diagnostics validation pass stops it being formatted anyway, but I think that might be unintentional.